README.md
EVE Online Python API
=====================
Python interface for EVE Online API.

Installation
------------

```bash
pip install git+https://github.com/a-tal/eveapi.git@forked
```


Example
-------

```python
import eveapi

api = eveapi.EVEAPIConnection()
auth = api.auth(keyID=API_KEY_ID, vCode=API_VER_CODE)

for character in auth.account.Characters():
    print(character.name)
```


For more examples see tests/test_api.py.


Testing
-------

Set the EVEAPI_TEST_KEYID and EVEAPI_TEST_VCODE environment variables, either
in your shell, in the setup.py, or in the tox.ini. You can still run some tests
without auth.

To test in your python environment:
```bash
$ python setup.py test
```

To test all supported pythons:

```bash
$ tox
```


CHANGELOG.md
Version: 1.4.1 - 05 May 2016
- implements caching. py3 compatible, tests with py.test

Version: 1.4 - 21 February 2015
- Moved package to folder, split one file to speparate files.

Version: 1.3.1 - 02 November 2014
- Fix problem with strings ending in spaces (this is not supposed to happen,
  but apparently tiancity thinks it is ok to bypass constraints)

Version: 1.3.0 - 27 May 2014
- Added set_user_agent() module-level function to set the User-Agent header
  to be used for any requests by the library. If this function is not used,
  a warning will be thrown for every API request.

Version: 1.2.9 - 14 September 2013
- Updated error handling: Raise an AuthenticationError in case
   the API returns HTTP Status Code 403 - Forbidden

Version: 1.2.8 - 9 August 2013
- the XML value cast function (_autocast) can now be changed globally to a
  custom one using the set_cast_func(func) module-level function.

Version: 1.2.7 - 3 September 2012
- Added get() method to Row object.

Version: 1.2.6 - 29 August 2012
- Added finer error handling + added setup.py to allow distributing eveapi
  through pypi.

Version: 1.2.5 - 1 August 2012
- Row objects now have __hasattr__ and __contains__ methods

Version: 1.2.4 - 12 April 2012
- API version of XML response now available as _meta.version

Version: 1.2.3 - 10 April 2012
- fix for tags of the form <tag attr=bla ... />

Version: 1.2.2 - 27 February 2012
- fix for the workaround in 1.2.1.

Version: 1.2.1 - 23 February 2012
- added workaround for row tags missing attributes that were defined
  in their rowset (this should fix ContractItems)

Version: 1.2.0 - 18 February 2012
- fix handling of empty XML tags.
- improved proxy support a bit.

Version: 1.1.9 - 2 September 2011
- added workaround for row tags with attributes that were not defined
  in their rowset (this should fix AssetList)

Version: 1.1.8 - 1 September 2011
- fix for inconsistent columns attribute in rowsets.

Version: 1.1.7 - 1 September 2011
- auth() method updated to work with the new authentication scheme.

Version: 1.1.6 - 27 May 2011
- Now supports composite keys for IndexRowsets.
- Fixed calls not working if a path was specified in the root url.

Version: 1.1.5 - 27 Januari 2011
- Now supports (and defaults to) HTTPS. Non-SSL proxies will still work by
  explicitly specifying http:// in the url.

Version: 1.1.4 - 1 December 2010
- Empty explicit CDATA tags are now properly handled.
- _autocast now receives the name of the variable it's trying to typecast,
  enabling custom/future casting functions to make smarter decisions.

Version: 1.1.3 - 6 November 2010
- Added support for anonymous CDATA inside row tags. This makes the body of
  mails in the rows of char/MailBodies available through the .data attribute.

Version: 1.1.2 - 2 July 2010
- Fixed __str__ on row objects to work properly with unicode strings.

Version: 1.1.1 - 10 Januari 2010
- Fixed bug that causes nested tags to not appear in rows of rowsets created
  from normal Elements. This should fix the corp.MemberSecurity method,
  which now returns all data for members. [jehed]

Version: 1.1.0 - 15 Januari 2009
- Added Select() method to Rowset class. Using it avoids the creation of
  temporary row instances, speeding up iteration considerably.
- Added ParseXML() function, which can be passed arbitrary API XML file or
  string objects.
- Added support for proxy servers. A proxy can be specified globally or
  per api connection instance. [suggestion by graalman]
- Some minor refactoring.
- Fixed deprecation warning when using Python 2.6.

Version: 1.0.7 - 14 November 2008
- Added workaround for rowsets that are missing the (required!) columns
  attribute. If missing, it will use the columns found in the first row.
  Note that this is will still break when expecting columns, if the rowset
  is empty. [Flux/Entity]

Version: 1.0.6 - 18 July 2008
- Enabled expat text buffering to avoid content breaking up. [BigWhale]

Version: 1.0.5 - 03 February 2008
- Added workaround to make broken XML responses (like the "row:name" bug in
  eve/CharacterID) work as intended.
- Bogus datestamps before the epoch in XML responses are now set to 0 to
  avoid breaking certain date/time functions. [Anathema Matou]

Version: 1.0.4 - 23 December 2007
- Changed _autocast() to use timegm() instead of mktime(). [Invisible Hand]
- Fixed missing attributes of elements inside rows. [Elandra Tenari]

Version: 1.0.3 - 13 December 2007
- Fixed keyless columns bugging out the parser (in CorporationSheet for ex.)

Version: 1.0.2 - 12 December 2007
- Fixed parser not working with indented XML.

Version: 1.0.1
- Some micro optimizations

Version: 1.0
- Initial release


tests/conftest.py
import os
import shutil
import tempfile

import pytest

import eveapi


@pytest.fixture
def file_cache(request):
    """Creates a clean cache dir for the FileCache instance."""

    tempdir = os.path.join(tempfile.gettempdir(), "eveapi")
    if os.path.exists(tempdir):
        shutil.rmtree(tempdir)
    os.makedirs(tempdir)
    return eveapi.cache.FileCache(cache_dir=tempdir)


@pytest.fixture
def api(request):
    return eveapi.EVEAPIConnection()


def _get_credential(env_var):
    credential = os.environ.get(env_var)
    if not credential:
        error = (
            "missing test credentials (set the environment variables "
            "EVEAPI_TEST_KEYID and EVEAPI_TEST_VCODE)"
        )
        pytest.xfail(error)
        pytest.fail(error)
    return credential


@pytest.fixture
def key_id(request):
    return _get_credential("EVEAPI_TEST_KEYID")


@pytest.fixture
def vcode(request):
    return _get_credential("EVEAPI_TEST_VCODE")


@pytest.fixture
def authenticated_api(api, key_id, vcode):
    return api.auth(keyID=key_id, vCode=vcode)


tests/test_api.py
# =============================================================================
# eveapi module demonstration script - Jamie van den Berge
# =============================================================================
#
# This file is in the Public Domain - Do with it as you please.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE
#
# ----------------------------------------------------------------------------

import mock
import pytest
from datetime import datetime

import eveapi


def test_alliance_list(api):
    """EXAMPLE 1: GETTING THE ALLIANCE LIST
    (and showing alliances with 1000 or more members)
    """

    # Let's get the list of alliances.
    # The API function we need to get the list is:
    #
    #    /eve/AllianceList.xml.aspx
    #
    # There is a 1:1 correspondence between folders/files and attributes on api
    # objects, so to call this particular function, we simply do this:
    result1 = api.eve.AllianceList()

    # This result contains a rowset object called "alliances". Rowsets are like
    # database tables and you can do various useful things with them. For now
    # we'll just iterate over it and display all alliances with more than 1000
    # members:

    assert result1.alliances, "no alliance results found"

    for alliance in result1.alliances:
        if alliance.memberCount >= 1000:
            print("{} <{}> has {:,d} members".format(
                alliance.name,
                alliance.shortName,
                alliance.memberCount,
            ))

    return result1.alliances


def test_wallet_balance(authenticated_api):
    """EXAMPLE 2: GETTING WALLET BALANCE OF ALL YOUR CHARACTERS"""

    # To get any info on character/corporation related stuff, we need to
    # acquire an authentication context. All API requests that require
    # authentication need to be called through this object. While it is
    # possible to call such API functions directly through the api object,
    # you would have to specify the userID and apiKey on every call. If you
    # are iterating over many accounts, that may actually be the better
    # option. However, for these examples we only  use one account, so this
    # is more convenient.

    # Now let's say you want to the wallet balance of all your characters.
    # The API function we need to get the characters on your account is:
    #
    #    /account/Characters.xml.aspx
    #
    # As in example 1, this simply means adding folder names as attributes
    # and calling the function named after the base page name:
    result2 = authenticated_api.account.Characters()

    # Some tracking for later examples.
    rich = 0
    rich_charID = 0

    # Now the best way to iterate over the characters on your account and show
    # the isk balance is probably this way:
    for character in result2.characters:
        wallet = authenticated_api.char.AccountBalance(
            characterID=character.characterID
        )
        isk = wallet.accounts[0].balance
        assert isinstance(isk, float)
        print("{} has {:,.2f} ISK.".format(character.name, isk))

        if isk > rich:
            rich = isk
            rich_charID = character.characterID

    return rich_charID


def test_failures(api):
    """EXAMPLE 3: WHEN STUFF GOES WRONG"""
    # Obviously you cannot assume an API call to succeed. There's a myriad of
    # things that can go wrong:
    #
    # - Connection error
    # - Server error
    # - Invalid parameters passed
    # - Hamsters died
    #
    # Therefor it is important to handle errors properly. eveapi will raise
    # an AttributeError if the requested function does not exist on the server
    # (ie. when it returns a 404), a RuntimeError on any other webserver error
    # (such as 500 Internal Server error).
    # On top of this, you can get any of the httplib (which eveapi uses) and
    # socket (which httplib uses) exceptions so you might want to catch those
    # as well.
    #

    with pytest.raises(eveapi.Error) as error:
        # Try calling account/Characters without authentication context
        api.account.Characters()

    assert error.value.code == 400
    assert error.value.args[0] == (
        "'/account/Characters.xml.aspx' request failed (Bad Request)"
    )


# -----------------------------------------------------------------------------
def test_character_sheet(authenticated_api):
    """EXAMPLE 4: GETTING CHARACTER SHEET INFORMATION"""

    # We grab ourselves a character context object.
    # Note that this is a convenience function that takes care of passing the
    # characterID=x parameter to every API call much like auth() does (in fact
    # it's exactly like that, apart from the fact it also automatically adds
    # the "/char" folder). Again, it is possible to use the API functions
    # directly from the api or auth context, but then you have to provide the
    # missing keywords on every call (characterID in this case).
    #
    # The victim we'll use is the last character on the account we used in
    # example 2.
    rich_charID = test_wallet_balance(authenticated_api)
    me = authenticated_api.character(rich_charID)

    # Now that we have a character context, we can display skills trained on
    # a character. First we have to get the skill tree. A real application
    # would cache this data; all objects returned by the api interface can be
    # pickled.
    skilltree = authenticated_api.eve.SkillTree()

    # Now we have to fetch the charactersheet.
    # Note that the call below is identical to:
    #
    #   acc.char.CharacterSheet(characterID=your_character_id)
    #
    # But, as explained above, the context ("me") we created automatically
    # takes care of adding the characterID parameter and /char folder attribute
    sheet = me.CharacterSheet()

    # This list should look familiar. They're the skillpoints at each level for
    # a rank 1 skill. We could use the formula, but this is much simpler :)
    sp = [0, 250, 1414, 8000, 45255, 256000]

    total_sp = 0
    total_skills = 0

    # Now the fun bit starts. We walk the skill tree, and for every group in
    # the tree...
    for g in skilltree.skillGroups:

        skills_trained_in_this_group = False

        # ... iterate over the skills in this group...
        for skill in g.skills:

            # see if we trained this skill by checking the character sheet
            trained = sheet.skills.Get(skill.typeID, False)
            if trained:
                # yep, we trained this skill.

                # print the group name if we haven't done so already
                if not skills_trained_in_this_group:
                    print(g.groupName)
                    skills_trained_in_this_group = True

                # and display some info about the skill!
                print("- {} Rank({}) - SP: {:,d}/{:,d} - Level: {}".format(
                    skill.typeName, skill.rank, trained.skillpoints,
                    (skill.rank * sp[trained.level]), trained.level)
                )
                total_skills += 1
                total_sp += trained.skillpoints

    # And to top it off, display totals.
    print("You currently have {} skills and {:,d} skill points".format(
        total_skills, total_sp)
    )


def test_rowsets(api):
    """EXAMPLE 5: USING ROWSETS"""

    # For this one we will use the result1 that contains the alliance list from
    # the first example.
    rowset = test_alliance_list(api)

    # Now, what if we want to sort the alliances by ticker name. We could
    # unpack all alliances into a list and then use python's sort(key=...) on
    # that list, but that's not efficient. The rowset objects support sorting
    # on columns directly:
    rowset.SortBy("shortName")

    # Note the use of Select() here. The Select method speeds up iterating over
    # large rowsets considerably as no temporary row instances are created.
    for ticker in rowset.Select("shortName"):
        assert ticker

    # The sort above modified the result inplace. There is another method,
    # called SortedBy, which returns a new rowset.
    sorted_by = rowset.SortedBy("allianceID", reverse=True)
    assert id(sorted_by) != id(rowset)

    # Another useful method of rowsets is IndexBy, which enables you to do
    # direct key lookups on columns. We already used this feature in example 3.
    # Indeed most rowsets returned are IndexRowsets already if the data has a
    # primary key attribute defined in the <rowset> tag in the XML data.
    #
    # IndexRowsets are efficient, they reference the data from the rowset they
    # were created from, and create an index mapping on top of it.
    #
    # Anyway, to create an index:
    alliances_by_ticker = rowset.IndexedBy("shortName")

    # Now use the Get() method to get a row directly.
    # Assumes ISD alliance exists. If it doesn't, we probably have bigger
    # problems than the unhandled exception here -_-
    assert alliances_by_ticker.Get("ISD")

    # You may specify a default to return in case the row wasn't found:
    assert alliances_by_ticker.Get("123456", 42) == 42

    # If no default was specified and you try to look up a key that does not
    # exist, an appropriate exception will be raised:
    with pytest.raises(KeyError):
        alliances_by_ticker.Get("123456")


def test_file_caching_data(file_cache):
    """EXAMPLE 6: CACHING DATA"""

    # Now try out the handler! Even though were initializing a new api object
    # here, a handler can be attached or removed from an existing one at any
    # time with its setcachehandler() method.
    cached_api = eveapi.EVEAPIConnection(cacheHandler=file_cache)

    assert len(cached_api._root._handler.cache) == 0

    cached_api.eve.SkillTree()
    assert len(cached_api._root._handler.cache) == 1

    # the second time it should be returning the cached version
    with mock.patch.object(eveapi.context, "httplib") as patched_httplib:
        cached_api.eve.SkillTree()

    assert len(cached_api._root._handler.cache) == 1
    assert patched_httplib.called is False


def test_transactions(file_cache, key_id, vcode):
    """EXAMPLE 7: TRANSACTION DATA
    (and doing more nifty stuff with rowsets)"""

    file_cached_auth_api = eveapi.EVEAPIConnection(
        cacheHandler=file_cache
    ).auth(keyID=key_id, vCode=vcode)

    rich_char_id = test_wallet_balance(file_cached_auth_api)
    rich_char = file_cached_auth_api.character(rich_char_id)
    journal = rich_char.WalletJournal()

    # Let's see how much we paid SCC in transaction tax in the first page
    # of data!

    # Righto, now we -could- sift through the rows and extract what we want,
    # but we can do it in a much more clever way using the GroupedBy method
    # of the rowset in the result. This creates a mapping that maps keys
    # to Rowsets of all rows with that key value in specified column.
    # These data structures are also quite efficient as the only extra data
    # created is the index and grouping.
    entriesByRefType = journal.transactions.GroupedBy("refTypeID")

    # Also note that we're using a hardcoded refTypeID of 54 here. You're
    # supposed to use .eve.RefTypes() though (however they are not likely
    # to be changed anyway so we can get away with it)
    # Note the use of Select() to speed things up here.
    amount = 0.0
    date = 0
    try:
        for result in entriesByRefType[54].Select("amount", "date"):
            tax_amount, date = result
            assert isinstance(date, datetime)
            amount -= tax_amount
    except KeyError:  # no taxes paid
        pass

    print("You paid a {:,.2f} ISK transaction tax since {}".format(
        amount,
        date,
    ))

    # You might also want to see how much a certain item yielded you recently.
    typeName = "Expanded Cargohold II"  # change this to something you sold.
    amount = 0.0
    date = 0

    wallet = rich_char.WalletTransactions()
    try:
        soldTx = wallet.transactions.GroupedBy("transactionType")["sell"]
        for row in soldTx.GroupedBy("typeName")[typeName]:
            amount += (row.quantity * row.price)
            date = row.transactionDateTime
            assert isinstance(date, datetime)
    except KeyError:  # has not sold any
        pass

    print("{} sales yielded {:,.2f} ISK since {}".format(
        typeName,
        amount,
        date,
    ))

    # this isn't really a test. but it does a bunch of stuff that can error


tests/test_package.py
import eveapi


def test_imports():
    should_exist = [
        "EVEAPIConnection",
        "Error",
        "AuthenticationError",
        "RequestError",
        "ServerError",
    ]
    top_level_objs = dir(eveapi)
    for obj in should_exist:
        assert obj in top_level_objs


eveapi/api.py
try:
    from urlparse import urlparse
except ImportError:
    from urllib.parse import urlparse

from eveapi.cache import MemoryCache
from eveapi.context import _RootContext

proxy = None
proxySSL = False


def set_user_agent(user_agent_string):
    """
    Sets a User-Agent for any requests sent by the library.
    """
    global USER_AGENT
    USER_AGENT = user_agent_string


def EVEAPIConnection(url="api.eveonline.com", cacheHandler=MemoryCache,
                     proxy=None, proxySSL=False):
    """
    Creates an API object through which you can call remote functions.

    The following optional arguments may be provided:

    url - root location of the EVEAPI server

    proxy - (host,port) specifying a proxy server through which to request
            the API pages. Specifying a proxy overrides default proxy.

    proxySSL - True if the proxy requires SSL, False otherwise.

    cacheHandler - an object which must support the following interface:

         retrieve(host, path, params)

             Called when eveapi wants to fetch a document.
             host is the address of the server, path is the full path to
             the requested document, and params is a dict containing the
             parameters passed to this api call (keyID, vCode, etc).
             The method MUST return one of the following types:

              None - if your cache did not contain this entry
              str/unicode - eveapi will parse this as XML
              Element - previously stored object as provided to store()
              file-like object - eveapi will read() XML from the stream.

         store(host, path, params, doc, obj)

             Called when eveapi wants you to cache this item.
             You can use obj to get the info about the object (cachedUntil
             and currentTime, etc) doc is the XML document the object
             was generated from. It's generally best to cache the XML, not
             the object, unless you pickle the object. Note that this method
             will only be called if you returned None in the retrieve() for
             this object.
    """

    if not url.startswith("http"):
        url = "https://" + url
    p = urlparse(url, "https")
    if p.path and p.path[-1] == "/":
        p.path = p.path[:-1]
    ctx = _RootContext(None, p.path, {}, {})
    ctx.setcachehandler(cacheHandler)
    ctx._scheme = p.scheme
    ctx._host = p.netloc
    ctx._proxy = proxy or globals()["proxy"]
    ctx._proxySSL = proxySSL or globals()["proxySSL"]
    return ctx


eveapi/cache.py
"""This used to be used in testing, might as well make it available."""


import os
import zlib
import pickle
from datetime import datetime


class _CacheBase(object):
    """Subclass this to implement storage. Do not use this directly.

    Implement `save`, `fetch` and `delete` methods in your subclass.
    """

    def __init__(self):
        self.cache = {}

    def retrieve(self, host, path, params):
        # eveapi asks if we have this request cached
        key = hash((host, path, frozenset(list(params.items()))))

        # see if we have the key in memory or fallback to implementation
        cached = self.cache.get(key, self.fetch(key))

        if cached is not None:
            content, until = cached
            # check if the cached doc is fresh enough
            if datetime.utcnow() < until:
                return content  # return the cached XML doc

            # it's stale. purge it from memory
            del self.cache[key]

            # and in our active implementation
            self.delete(key)

    def store(self, host, path, params, doc, obj):
        # eveapi is asking us to cache an item
        key = hash((host, path, frozenset(list(params.items()))))

        cached_for = obj.cachedUntil - obj.currentTime
        if cached_for:
            # store in memory
            self.cache[key] = (doc, obj.cachedUntil)

            # store in implementation
            self.save(key, doc, obj.cachedUntil)


class MemoryCache(_CacheBase):
    """Stores returns in memory for the intended timeframe."""

    def save(self, key, content, until):
        """This is passed the key name, the content and a datetime until."""
        pass

    def fetch(self, key):
        """This should return a tuple of the content, and datetime until.
        None if not found."""
        pass

    def delete(self, key):
        """This should delete the content based on its key."""
        pass


class FileCache(_CacheBase):
    """
    Note: this is an example handler to demonstrate how to use them.
    a -real- handler should probably be thread-safe and handle errors
    properly (and perhaps use a better hashing scheme).

    Initalize with a directory filepath to use for storing .cache files.
    """

    def __init__(self, cache_dir):
        if not os.path.isdir(cache_dir):
            raise OSError("{} is not a directory".format(cache_dir))
        self.cache_dir = cache_dir
        super(FileCache, self).__init__()

    def _get_cache_file(self, key):
        return os.path.join(self.cache_dir, "{}.cache".format(key))

    def fetch(self, key):
        """Fetches (content, until) from our cache folder."""

        cache_file = self._get_cache_file(key)
        if os.path.exists(cache_file):
            with open(cache_file, "rb") as opencache:
                return pickle.loads(zlib.decompress(opencache.read()))

    def delete(self, key):
        """Removes the .cache file from disk."""

        try:
            os.remove(self._get_cache_file(key))
        except:
            pass

    def save(self, key, content, until):
        """Saves (content, until) to our cache folder."""

        cache_file = os.path.join(self.cache_dir, "{}.cache".format(key))
        with open(cache_file, "wb") as opencache:
            opencache.write(zlib.compress(pickle.dumps((content, until), -1)))


eveapi/context.py
try:
    import httplib
except ImportError:
    import http.client as httplib

try:
    from urllib import urlencode
except ImportError:
    from urllib.parse import urlencode
import warnings

from eveapi import DEFAULT_UA, USER_AGENT
from eveapi.exceptions import Error, ServerError, AuthenticationError
from eveapi.parser import _ParseXML


class _Context(object):

    def __init__(self, root, path, parentDict, newKeywords=None):
        self._root = root or self
        self._path = path
        if newKeywords:
            if parentDict:
                self.parameters = parentDict.copy()
            else:
                self.parameters = {}
            self.parameters.update(newKeywords)
        else:
            self.parameters = parentDict or {}

    def context(self, *args, **kw):
        if kw or args:
            path = self._path
            if args:
                path += "/" + "/".join(args)
            return self.__class__(self._root, path, self.parameters, kw)
        else:
            return self

    def __getattr__(self, this):
        # perform arcane attribute majick trick
        return _Context(self._root, self._path + "/" + this, self.parameters)

    def __call__(self, **kw):
        if kw:
            # specified keywords override contextual ones
            for k, v in self.parameters.items():
                if k not in kw:
                    kw[k] = v
        else:
            # no keywords provided, just update with contextual ones.
            kw.update(self.parameters)

        # now let the root context handle it further
        return self._root(self._path, **kw)


class _AuthContext(_Context):

    def character(self, characterID):
        # returns a copy of this connection object but for every call made
        # through it, it will add the folder "/char" to the url, and the
        # characterID to the parameters passed.
        return _Context(self._root, self._path + "/char", self.parameters,
                        {"characterID": characterID})

    def corporation(self, characterID):
        # same as character except for the folder "/corp"
        return _Context(self._root, self._path + "/corp", self.parameters,
                        {"characterID": characterID})


class _RootContext(_Context):

    def auth(self, **kw):
        if len(kw) == 2 and (("keyID" in kw and "vCode" in kw) or
           ("userID" in kw and "apiKey" in kw)):
            return _AuthContext(self._root, self._path, self.parameters, kw)
        raise ValueError("Must specify keyID and vCode")

    def setcachehandler(self, handler):
        if isinstance(handler, type):
            # late initialization of cache implementation
            handler = handler()
        self._root._handler = handler

    def __call__(self, path, **kw):
        # convert list type arguments to something the API likes
        for k, v in kw.items():
            if isinstance(v, (list, tuple, dict)):
                kw[k] = ','.join(map(str, list(v)))

        cache = self._root._handler

        # now send the request
        path += ".xml.aspx"

        if cache:
            response = cache.retrieve(self._host, path, kw)
        else:
            response = None

        if response is None:
            if not USER_AGENT:
                warnings.warn(
                    "No User-Agent set! Please use the set_user_agent() "
                    "module-level function before accessing the EVE API.",
                    stacklevel=3,
                )

            if self._proxy is None:
                req = path
                if self._scheme == "https":
                    conn = httplib.HTTPSConnection(self._host)
                else:
                    conn = httplib.HTTPConnection(self._host)
            else:
                req = '{}://{}{}'.format(self._scheme, self._host, path)
                if self._proxySSL:
                    conn = httplib.HTTPSConnection(*self._proxy)
                else:
                    conn = httplib.HTTPConnection(*self._proxy)

            if kw:
                conn.request("POST", req, urlencode(kw), {
                    "Content-type": "application/x-www-form-urlencoded",
                    "User-Agent": USER_AGENT or DEFAULT_UA,
                })
            else:
                conn.request("GET", req, "", {
                    "User-Agent": USER_AGENT or DEFAULT_UA
                })

            response = conn.getresponse()
            if response.status != 200:
                if response.status == httplib.NOT_FOUND:
                    raise AttributeError(
                        "'%s' not available (404 Not Found)" % path
                    )
                elif response.status == httplib.FORBIDDEN:
                    raise AuthenticationError(response.status,
                                              'HTTP 403 - Forbidden')
                else:
                    raise ServerError(
                        response.status,
                        "'%s' request failed (%s)" % (path, response.reason),
                    )

            if cache:
                store = True
                response = response.read()
            else:
                store = False
        else:
            store = False

        fallback = cache and getattr(cache, "retrieve_fallback", False)
        if fallback:
            # implementor is handling fallbacks...
            try:
                return _ParseXML(
                    response,
                    True,
                    store and (lambda obj: cache.store(
                        self._host,
                        path,
                        kw,
                        response,
                        obj,
                    )),
                )
            except Error as e:
                response = fallback(self._host, path, kw, reason=e)
                if response is not None:
                    return response
                raise
        else:
            # implementor is not handling fallbacks...
            return _ParseXML(
                response,
                True,
                store and (lambda obj: cache.store(
                    self._host,
                    path,
                    kw,
                    response,
                    obj,
                )),
            )
